server/consumer.go

```diff
@@ -2328,17 +2328,22 @@ func (o *consumer) isFiltered() bool {
 
 // Check if we need an ack for this store seq.
 // This is called for interest based retention streams to remove messages.
-func (o *consumer) needAck(sseq uint64) bool {
+func (o *consumer) needAck(sseq uint64, subj string) bool {
 	var needAck bool
 	var asflr, osseq uint64
 	var pending map[uint64]*Pending
 	o.mu.RLock()
 
-	// Check first if we are filtered, and if so check if this is even applicable to us.
+	// Check if we are filtered, and if so check if this is even applicable to us.
 	if o.isFiltered() && o.mset != nil {
-		var svp StoreMsg
-		sm, err := o.mset.store.LoadMsg(sseq, &svp)
-		if err != nil || !o.isFilteredMatch(sm.subj) {
+		if subj == _EMPTY_ {
+			var svp StoreMsg
+			if _, err := o.mset.store.LoadMsg(sseq, &svp); err != nil {
+				return false
```

Comment:
I guess, `o.mu.RUnlock()` is missed here (before `return`)?

Reply:
Looks like it — https://github.com/nats-io/nats-server/pull/3555